src/components/app/index.tsx

```diff
@@ -0,0 +1,112 @@
+import { h, Component } from 'preact';
+import { updater, toggle, When } from '../../lib/util';
+import Fab from '../fab';
+import Header from '../header';
+// import Drawer from 'async!../drawer';
+const Drawer = require('async!../drawer').default;
+import Home from '../home';
+import * as style from './style.scss';
+
+type Props = {
+	url?: String
+}
+
+type FileObj = {
+	id: any,
+	data: any,
+	error: Error | DOMError | String,
+	file: File,
+	loading: Boolean
+};
+
+type State = {
+	showDrawer: Boolean,
+	showFab: Boolean,
+	files: FileObj[]
+};
+
+let counter = 0;
+
+export default class App extends Component<Props, State> {
+	state: State = {
+		showDrawer: false,
+		showFab: true,
+		files: []
+	};
+
+	loadFile = (file: File) => {
+		let fileObj = {
+			id: ++counter,
+			file,
+			error: null,
+			loading: true,
+			data: null
+		};
+
+		this.setState({
+			files: [fileObj]
+		});
+
+		let fr = new FileReader();
+		// fr.readAsArrayBuffer();
+		fr.onerror = () => {
+			let files = this.state.files.slice();
+			files.splice(0, files.indexOf(fileObj), {
+				...fileObj,
+				error: fr.error,
+				loading: false
+			});
+			this.setState({ files });
+		};
+		fr.onloadend = () => {
+			let files = this.state.files.slice();
+			files.splice(0, files.indexOf(fileObj), {
+				...fileObj,
+				data: fr.result,
+				loading: false
+			});
+			this.setState({ files });
+		};
+		fr.readAsDataURL(file);
+	};
+
+	enableDrawer = false;
+
+	openDrawer = updater(this, 'showDrawer', true);
```

Comment:
These should be proper methods. We can use TypeScript decorators to tie in stuff like `updater()`

Reply:
I'll try making a decorator. Personally I've been gravitating towards using a "store" to house data + mutators/actions.  I also threw some ideas around with @davideast last week for simplifying state+selectors, and he published something called [selectrify](https://github.com/davideast/selectrify).